src/coreclr/System.Private.CoreLib/src/System/Array.CoreCLR.cs

```diff
@@ -249,147 +249,152 @@ private static unsafe void CopyImplPrimitiveWiden(Array sourceArray, int sourceI
-                            default:
-                                Debug.Fail("Array.Copy from U4 to another type hit unsupported widening conversion"); break;
-                        }
-                        break;
-
-                    case CorElementType.ELEMENT_TYPE_I4:
-                        switch (destElType)
-                        {
-                            case CorElementType.ELEMENT_TYPE_I8:
-                                Unsafe.As<byte, long>(ref destElement) = Unsafe.As<byte, int>(ref srcElement); break;
-                            case CorElementType.ELEMENT_TYPE_R4:
-                                Unsafe.As<byte, float>(ref destElement) = Unsafe.As<byte, int>(ref srcElement); break;
-                            case CorElementType.ELEMENT_TYPE_R8:
-                                Unsafe.As<byte, double>(ref destElement) = Unsafe.As<byte, int>(ref srcElement); break;
-                            default:
-                                Debug.Fail("Array.Copy from I4 to another type hit unsupported widening conversion"); break;
-                        }
-                        break;
-
-                    case CorElementType.ELEMENT_TYPE_U8:
-                        switch (destElType)
-                        {
-                            case CorElementType.ELEMENT_TYPE_R4:
-                                Unsafe.As<byte, float>(ref destElement) = Unsafe.As<byte, ulong>(ref srcElement); break;
-                            case CorElementType.ELEMENT_TYPE_R8:
-                                Unsafe.As<byte, double>(ref destElement) = Unsafe.As<byte, ulong>(ref srcElement); break;
-                            default:
-                                Debug.Fail("Array.Copy from U8 to another type hit unsupported widening conversion"); break;
-                        }
-                        break;
-
-                    case CorElementType.ELEMENT_TYPE_I8:
-                        switch (destElType)
-                        {
-                            case CorElementType.ELEMENT_TYPE_R4:
-                                Unsafe.As<byte, float>(ref destElement) = Unsafe.As<byte, long>(ref srcElement); break;
-                            case CorElementType.ELEMENT_TYPE_R8:
-                                Unsafe.As<byte, double>(ref destElement) = Unsafe.As<byte, long>(ref srcElement); break;
-                            default:
-                                Debug.Fail("Array.Copy from I8 to another type hit unsupported widening conversion"); break;
-                        }
-                        break;
-
-                    case CorElementType.ELEMENT_TYPE_R4:
-                        Debug.Assert(destElType == CorElementType.ELEMENT_TYPE_R8);
-                        Unsafe.As<byte, double>(ref destElement) = Unsafe.As<byte, float>(ref srcElement); break;
+                PrimitiveWiden(ref srcElement, ref destElement, srcElType, destElType);
```

Comment:
```suggestion
                PrimitiveWiden(ref srcData, ref data, srcElType, destElType);
                srcData = ref Unsafe.AddByteOffset(ref srcData, srcElSize);
                data= ref Unsafe.AddByteOffset(ref data, destElSize);
```
Match naot - better perf, avoids multiplications per iteration.